internal/pkg/gateway/api.go

```diff
@@ -173,183 +169,190 @@ func (gs *Server) Endorse(ctx context.Context, request *gp.EndorseRequest) (*gp.
-						gs.registry.removeEndorser(firstEndorser)
-					}
-					firstEndorser = plan.nextPeerInGroup(firstEndorser)
-					firstResponse = nil
+	}
+
+	for plan.completed == nil {
+		// loop through the layouts until one gets satisfied
+		endorsers := plan.endorsers()
+		if endorsers == nil {
+			// no more layouts
+			break
+		}
+		// send to all the endorsers
+		waitCh := make(chan bool, len(endorsers))
+		for _, e := range endorsers {
+			go func(e *endorser) {
+				success := false
+				for !success && e != nil {
+					e, success = gs.processProposal(ctx, plan, e, signedProposal, logger)
 				}
-			}()
+				waitCh <- true
+			}(e)
+		}
+		for i := 0; i < len(endorsers); i++ {
 			select {
-			case <-done:
+			case <-waitCh:
 				// Endorser completed normally
 			case <-ctx.Done():
-				// Overall endorsement timeout expired
-				logger.Warnw("Endorse call timed out while collecting first endorsement", "channel", request.ChannelId, "txID", request.TransactionId)
-				return nil, newRpcError(codes.DeadlineExceeded, "endorsement timeout expired while collecting first endorsement")
+				logger.Warnw("Endorse call timed out while collecting endorsements", "numEndorsers", len(endorsers))
+				return nil, newRpcError(codes.DeadlineExceeded, "endorsement timeout expired while collecting endorsements")
 			}
 		}
-		if firstEndorser == nil || firstResponse == nil {
-			return nil, newRpcError(codes.Aborted, "failed to endorse transaction, see attached details for more info", errDetails...)
-		}
 
-		// 3. Extract ChaincodeInterest and SBE policies
-		// The chaincode interest could be nil for legacy peers and for chaincode functions that don't produce a read-write set
-		interest := firstResponse.Interest
-		if len(interest.GetChaincodes()) == 0 {
-			interest = defaultInterest
-		}
+	}
 
-		// 4. If transient data is involved, then we need to ensure that discovery only returns orgs which own the collections involved.
-		// Do this by setting NoPrivateReads to false on each collection
-		if hasTransientData {
-			for _, call := range interest.GetChaincodes() {
-				call.NoPrivateReads = false
+	if plan.completed == nil {
+		return nil, newRpcError(codes.Aborted, "failed to collect enough transaction endorsements, see attached details for more info", plan.errorDetails...)
+	}
+
+	action = &peer.ChaincodeEndorsedAction{ProposalResponsePayload: plan.responsePayload, Endorsements: uniqueEndorsements(plan.completed.endorsements)}
+
+	preparedTransaction, err := prepareTransaction(header, payload, action)
+	if err != nil {
+		return nil, status.Errorf(codes.Aborted, "failed to assemble transaction: %s", err)
+	}
+
+	return &gp.EndorseResponse{PreparedTransaction: preparedTransaction}, nil
+}
+
+type ppResponse struct {
+	response *peer.ProposalResponse
+	err      error
+}
+
+func (gs *Server) processProposal(ctx context.Context, plan *plan, endorser *endorser, signedProposal *peer.SignedProposal, logger *flogging.FabricLogger) (*endorser, bool) {
```

Comment:
For these non-trivial functions, it helps to have a function comment explaining what the function does.